### ISSUE TITLE
Disable color output for the single instance

### ISSUE THREAD

**Issue by ai** [jorgebucaran/colorette#42]
Hi. Can I have color output in one part of the app and non-color in other? Something like:

```js
const c = colorette.instance({ enabled: opts.colors })
c.red("Error")
```

**Comment by jorgebucaran** [comment]
Nope, there's no API for that at the moment.

**Comment by snitin315** [comment]
@jorgebucaran How can I disable colors for GitHub CI?

**Comment by jorgebucaran** [comment]
@snitin315 The simplest way to disable color output is by using the [`NO_COLOR`](https://no-color.org/) env variable, e.g.

```console
NO_COLOR= node my_program.js
```

**Comment by snitin315** [comment]
@jorgebucaran I'm facing an error. 
![Screenshot at 2020-07-27 17-33-59](https://user-images.githubusercontent.com/46647141/88539788-80bafd80-d02f-11ea-9beb-90c3f30eac11.png)

**Comment by jorgebucaran** [comment]
Are you on Windows? And if so, do you know what is the Windows equivalent of `FOO= my_program`?

That doesn't seem like a colorette specific issue to me.

**Comment by snitin315** [comment]
We are migrating our project from `chalk` to `colorette`, All colored value tests are passing fine on Mac and Linux but some are failing for windows.
for example if I do expect('red color').toEqual('red('red color')'); // works fine on Mac n Linux but fails on windows.

**Comment by jorgebucaran** [comment]
@snitin315 Would you create a new issue for this? This issue is for another discussion. Thanks.

**Comment by jorgebucaran** [comment]
Sure, we could look into this, but I want to understand why this is needed to start with. Care to explain what's the problem with pinojs/pino-pretty#220? I still don't understand what's happening.

**Comment by kibertoad** [comment]
Because it leaks state. Imagine I already use Colorette in my project, and then I add pino-pretty, and this transitive dependency suddenly affects my main project and turns on colourizing globally even though it was off before.

**Comment by jorgebucaran** [comment]
You are describing the problem of "global state" (evil, I know), but is that what's actually happening in pinojs/pino-pretty#220? Do you have any guesses as to where the unwanted mutation is? Could it be pino-pretty changing `.enabled` or is it that user's code?

**Comment by kibertoad** [comment]
@jorgebucaran Original problem is a difference in behaviour between chalk and colourette. It was possible for pino-pretty to explicitly force Chalk instance to colourize, bypassing all checks. Colourette requires global state changes to achieve the same.

**Comment by kibertoad** [comment]
Here is the change that broke the support:
![image](https://user-images.githubusercontent.com/1847934/131640047-988bb473-eba1-4a36-bc41-7dd9d18297fe.png)

**Comment by jorgebucaran** [comment]
Oh, okay, so pino-pretty always colorizes regardless of terminal support (I mean that's what you want), correct?

**Comment by kibertoad** [comment]
@jorgebucaran To be precise, it works like this:
1) It tells Chalk to colourize regardless of terminal support;
2) Then it asks Chalk if there is terminal support. If it does, or if user explicitly enabled colourizing, it enables colourizing internally (pino-pretty will attempt to colourize);

Hence currently default colourizing behaviour is correct, but behaviour for when user has explicitly enabled colourizing is not, because it will also additionally check if there is terminal support for colourizing, and if not, ignore explicitly set colourizing flag.

**Comment by jorgebucaran** [comment]
So, the user actually enabled color, but that didn't actually set Colorette's `.enabled` to anything, right?

Can you tell me where in pino-pretty is defined the behavior for when the user has explicitly enabled color? Want to see how it was implemented.

**Comment by kibertoad** [comment]
@jorgebucaran https://github.com/pinojs/pino-pretty/pull/218/commits/bb1afa502173915168e4d2f761a7664a2e90ffd4 this was the place.

**Comment by kibertoad** [comment]
Another relevant part:
![image](https://user-images.githubusercontent.com/1847934/131838956-a778ebe4-d2e5-4cc4-aec9-c8acb8d10f7a.png)

Note how it default to what Chalk or Colourette tells is supported, but now if user deviates from default, it is not passed back to Colourette. Previously it didn't matter as chalk part was always enabled.

**Comment by kibertoad** [comment]
@jorgebucaran opts.colorize either defaults to `coloretteOptions.enabled`, or can be forced to `true` is user passes `-c`

**Comment by kibertoad** [comment]
Main difference is that previously Chalk part was always enabled, so it was due to pino-pretty part either being enabled or not. Currently even if pino-pretty part is enabled, if Colorette thinks terminal doesn't support it, nothing will be colourized.

**Comment by jorgebucaran** [comment]
So, it serves the same purpose as Colorette's `.enabled`. Why can't we just do this again?

```js
coloretteOptions.enabled = opts.colorizer
```

**Comment by kibertoad** [comment]
@jorgebucaran Because see the very beginning of this discussion. This leaks global state and affects everything else that might be using Colorette as well, e. g. main application using pino-pretty which may want to have very different configuration. This is a no-go for pino-pretty.

**Comment by jorgebucaran** [comment]
> This leaks global state and affects everything else that might be using Colorette as well...

This is a bit vague. Could you specifically point out what would go wrong? Could you describe a scenario where someone would be affected or experience unexpected behavior if we did `coloretteOptions.enabled = opts.colorizer`. That would clarify everything for me.


> main application ... may want to have a very different configuration

I suppose this means either forcing color or disabling color. Just to be sure.

> Chalk avoids the problem by encapsulating colourization instances, ...

No worries, I got this part.

**Comment by kibertoad** [comment]
E. g. I am writing a CLI tool enterpriseCLI. I directly depend on Colorette, I use it for coloring my output. Within enterpriseCLI I set global value for `coloretteOptions.enabled` to disabled because I don't want colour for some reason in some particular environment. Then I introduce pino-pretty as a dependency. Pino-pretty checks colorette parameter, sees that this environment supports colours, rewrites global value for `coloretteOptions.enabled` and leaks this parameter to my enterpriseCLI, colouring my output there as well. Problem will become worse as Colorette becomes more popular and is used in larger and larger amount of transitive dependencies, all of them will just start affecting each other.

**Comment by kibertoad** [comment]
It also works in reverse - if my environment doesn't explicitly support colours and I had to force-enable it in colorette it from my enterpriseCLI, I will also have to force-enable it in pino-pretty, otherwise pino-pretty will disable coloring for my enterpriseCLI as well.

**Comment by jorgebucaran** [comment]
Thank you, @kibertoad! I never considered this use case.  

As for the API. Would this work for you?

```js
import colors from "colorette"

// Enable color explicitly

const { red } = color(true)

// Forcefully disable color

const { red } = colors(false)

// or let me auto-detect it for you

const { red } = colors()
```

**Comment by jorgebucaran** [comment]
Thank you, @kibertoad! I never considered this use case.  

Would this API work for you?

```js
import colors from "colorette"

// Enable color explicitly

const { red } = color(true)

// Forcefully disable color

const { red } = colors(false)

// or let me auto-detect it for you

const { red } = colors()
```

**Comment by jorgebucaran** [comment]
Thank you, @kibertoad! I never considered this use case.  

Would this API work for you?

```js
import getColors from "colorette"

// Enable color explicitly

const { red } = getColors(true)

// Forcefully disable color

const { red } = getColors(false)

// or let me auto-detect it for you

const { red } = getColors()
```

**Comment by kibertoad** [comment]
@jorgebucaran perfect!

**Comment by jorgebucaran** [comment]
Great! Let me think it over a bit and if we don't come with anything better, we'll go with that.

**Comment by jorgebucaran** [comment]
Here's another idea! How about leaving it to the caller to wrap colors in any way they want?

```js
import * as colors from "colorette"

// ...

const { red, green, blue } = enterpriseOptions.enabled
  ? colors
  : Object.fromEntries(Object.entries(colors).map(([key]) => [key, c => c]))
```

> For convenience, we could export a new `none` function equivalent to `(c => c)` too.


You could use `colors.enabled` to set `enterpriseOptions.enabled` as well.

**Comment by kibertoad** [comment]
`enterpriseOptions.enabled` is this just some internal object, e. g.
```
const enterpriseOptions = {
  enabled: true,
}
```
?

What is the advantage of this approach? Sounds like same thing with extra steps, hence worse developer experience, because everyone would have to copy-paste same solution instead of having library provide it out-of-the-box.

**Comment by jorgebucaran** [comment]
`enterpriseOptions.enabled` would be `opts.colorizer` in pino-pretty, no? You can use `colors.enabled` too. It's up to you.

The advantage is keeping Colorette simple and give you the flexibility to wrap colors any way you want. 

The alternative is giving you that specific function:

```js
import colorette, { enabled } from "colorette"

// ...

const { red, green, blue } = colorette({ enabled: enterpriseOptions.enabled })
```

The disadvantage is that we have a default export now. Yes, DX is better, but I personally don't mind being explicit:

```js
import * as colors from "colorette"

// ...

const { red, green, blue } = enterpriseOptions.enabled
  ? colors
  : Object.fromEntries(Object.entries(colors).map(([key]) => [key, c => c]))
```

What do you think?

**Comment by kibertoad** [comment]
Explicit = boilerplate in this case, which I'm not a fan of.
Personally I don't see a problem with doing a semver major to replace default export with a named one (I generally much prefer named imports, but obviously that is a matter of taste).
Since you are the creator of the package, it is your call how this package should behave. If that particular pattern is documented so that everyone can simply copy-paste it in their codebase, it's not terrible :).

**Comment by kibertoad** [comment]
@jorgebucaran How would your proposed solution work with compatibility detection logic? Does it completely go away and if user is calling `colors.red`, then it will be always drawn, regardless of compatibility detection outcome?

**Comment by jorgebucaran** [comment]
> The more I think about it, the less I like this particular approach overall.

There are two approaches, which one do you mean?

> How would your proposed solution work with compatibility detection logic? 

`red()` would always output red. `enabled` would be true or false as usual, but it would no longer be a property, just a const.

**Comment by kibertoad** [comment]
@jorgebucaran The one when the whole burden of controlling when to draw lies on the shoulder of a developer (second option). Developer experience of that is really questionable, personally I would definitely just go for Chalk if those were my options.

```
import getColors from "colorette"

// Enable color explicitly

const { red } = getColors(true)

// Forcefully disable color

const { red } = getColors(false)

// or let me auto-detect it for you

const { red } = getColors()
```

This one works great, although I would probably still prefer to have it as a named export, so that { getColors, isColorSupported } could be done.

**Comment by jorgebucaran** [comment]
I like named exports too, but I could probably go with this:

```js
import c, { enabled } from "colorette"

const { red, green, blue } = c({ enabled })
```

We could even get rid of automatic color support, but I'd really prefer to keep that.

**Comment by jorgebucaran** [comment]
`supportsColor` might be good too.

**Comment by kibertoad** [comment]
@jorgebucaran

`import c, { enabled } from "colorette"

const { red, green, blue } = c({ enabled })`

Yupp, that looks great; defaulting to `enabled` when no param is passed, as per original suggestion, also sounds like a good idea.

**Comment by kibertoad** [comment]
Would you like to implement it yourself, or should I give it a shot?

**Comment by jorgebucaran** [comment]
Yeah, let's keep auto-detection built-in. Let me take care of this one! 👌

**Comment by jorgebucaran** [comment]
@kibertoad Just pushed the new API to the [`instance`](https://github.com/jorgebucaran/colorette/tree/instance) branch. It works as described in https://github.com/jorgebucaran/colorette/issues/42#issuecomment-912995002, give or take.

```js
import colorette, { isColorSupported } from "colorette"

if (!isColorSupported) throw new Error("Color not supported!")

// ...

const { blue } = colorette({ enableColor: myOptions.forceColor })

console.log(blue("I'm blue")) // => \x1B[34mI'm blue\x1B[39m
```

Now, before going forward with this, I have an alternative API I'd like to submit for consideration as well:

```js
import { blue, bold, underline, toColor } from "colorette"

console.log(
  toColor(blue("I'm blue"), { color: true })
)
```

You could wrap `toColor` to create a colorizer that's always enabled/disabled (or we could give it to you):

```js
import { blue, bold, underline, createColorizer } from "colorette"

const colorize = createColorizer({ color: true })

console.log(
  colorize(blue(underline(bold("I'm blue da ba dee da ba da"))))
)
```

Thoughts?

**Comment by kibertoad** [comment]
@jorgebucaran Do I understand correctly that the main advantage of the alternate approach is being able to chain different styles? 

`colorizeCLI(blue('text'))` looks less convenient than `blueCli('text')`, so if not for chaining, first one still looks like a better option. `toColor(blue('text'))` is definitely harder to read.

**Comment by jorgebucaran** [comment]
You should be able to chain or nest styles either way. The alternative is somewhat "purer". I also find it more natural to import the color functions rather than create them via a function.

While I prefer the alternative, I'm fine either way.

**Comment by kibertoad** [comment]
@jorgebucaran I'm just looking from developer's perspective, and as a developer I want as little boilerplate as possible. Using two functions where I could use just one to the same effect feels like a step backwards.

Maybe default instance could be created out-of-the-box, and color functions exported for it, so that those who do not need new API could just keep using them?

**Comment by jorgebucaran** [comment]
My position is far less extreme than yours. I wouldn't mind some boilerplate in exchange for simplicity and a bit more purity. 

Your suggestion could work, let me think about it too.

**Comment by jorgebucaran** [comment]
Just pushed the new documentation. Would you mind having a look at the [API section](https://github.com/jorgebucaran/colorette/tree/instance#api) and tell me if it makes sense to you or whether you'd do it differently? 🙏

**Comment by kibertoad** [comment]
@jorgebucaran Looks perfect!

**Comment by jorgebucaran** [comment]
Haha, thanks. 😄

**Comment by jorgebucaran** [comment]
@kibertoad I'm taking your suggestion to export individual colors (always enabled) in addition to the new API.

Now, did you notice the proposed `createColors({ colorize: boolean })` won't let you toggle color dynamically?

A different API like this would:

```js
import { Colorette } from "colorette"

const { blue, options } = new Colorette({ colorize: true })

console.log(c.blue("Blue"))  // => \x1B[34mBlue\x1B[39m

options.enabled = false

console.log(c.blue("Blue"))  // => Blue
```

**Comment by kibertoad** [comment]
@jorgebucaran I don't think this would be very useful. If someone needs to dynamically change configuration, they could simply create a new Colorette instance with a different configuration. Immutable instances are cleaner.

**Comment by jorgebucaran** [comment]
It's just like Chalk's `new Chalk()`, so maybe it can be useful, plus it would be consistent with Colorette's current `.enabled` property that lets you toggle color on or off as needed.

Still, do you reckon this solves https://github.com/pinojs/pino-pretty/issues/220?

**Comment by jorgebucaran** [comment]
FYI, here is the class-based implementation:

```js
export class Colorette {
  constructor({ colorize } = { colorize: isColorSupported }) {
    this.options = { colorize }

    Object.keys(colors).forEach((key) => {
      const color = colors[key]
      this[key] = (s) => (this.options.colorize ? color(s) : s)
    })
  }
}
```

and here's the factory-style implementation:

```js
export const createColors = (options = { colorize: isColorSupported }) => ({
  options,
  ...Object.entries(colors).reduce((colorMap, [key, color]) => ({
    ...colorMap,
    [key]: (s) => (options.colorize ? color(s) : s),
  })),
})
```

**Comment by kibertoad** [comment]
@jorgebucaran Yeah, either solution would resolve that issue.
No strong preference on class vs factory implementation, performance is also mostly identical on properly warmed up instance.
If you feel like having mutable options make sense - sure, but it is different from `new Chalk` specifically because that creates new immutable instance, does not mutate existing one

**Comment by jorgebucaran** [comment]
> @kibertoad  it is different from new Chalk specifically because that creates new immutable instance, does not mutate existing one

I must have gotten it all wrong. So, `new Chalk` won't let you change whether the output will be colorized?

**Comment by jorgebucaran** [comment]
> @kibertoad  it is different from new Chalk specifically because that creates new immutable instance, does not mutate existing one

I must have gotten it all wrong. So, `new Chalk` won't let you change whether the output will be colorized once created?

**Comment by kibertoad** [comment]
@jorgebucaran Only within that new Chalk instance.

**Comment by jorgebucaran** [comment]
> Only within that new Chalk instance 

Sounds like it will let you dynamically toggle color on or off, which is what I am suggesting [here](https://github.com/jorgebucaran/colorette/issues/42#issuecomment-915862837). Otherwise, how is `new Colorette` (or `createColors()`) different from `new Chalk`?

**Comment by kibertoad** [comment]
`new Colorette` is exactly the same as `new Chalk`. Chalk doesn't have equivalent for `{ options }; options.colorize = false` though.

**Comment by kibertoad** [comment]
@jorgebucaran What's the plan to proceed?

**Comment by jorgebucaran** [comment]
I feel `new Colorette` is more idiomatic without having you remember a function name (other than the name of the library itself). I should be able to finish the work and publish 2.0 in a few days.

I think this is not currently blocking `pino-pretty`, though, correct?

**Comment by jorgebucaran** [comment]
I feel `new Colorette` is more idiomatic without having you remember a new function name (other than the name of the library itself). I should be able to finish the work and publish 2.0 in a few days.

I think this is not currently blocking `pino-pretty`, though, right?

**Comment by kibertoad** [comment]
It is blocking. There are still regressions being reported with people losing colouring in pino-pretty.

**Comment by jorgebucaran** [comment]
Now available in `2.0.0`.

```js
const { createColors } from "colorette"

const { blue } = createColors(/* @param { useColor: boolean } */)

console.log(blue("2.0.0"))
```

[See the upgrading guide for details](https://github.com/jorgebucaran/colorette/issues/70).

**Comment by kibertoad** [comment]
Woohoo!
